### PR TITLE
support urls as ADD sources

### PR DESCRIPTION
**Pull request by diunko**

### CODE REVIEW

File: src/build/copy.go [hunk 1]
```diff
@@ -259,11 +305,28 @@ func listFiles(srcPath string, includes, excludes []string) ([]*uploadFile, erro
 		return nil, err
 	}
 
+	// XXX How should we get this guy here?
+	urlCache = b.urlCache
```
Comment (diunko) [inline src/build/copy.go]: @ybogdanov @romank87 any ideas, how should we bring this guy here?

**Comment by diunko** [comment]
@ybogdanov @romank87 PTAL

File: src/build/urlfetcher.go [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+func (info *URLInfo) getInfoFileName() (fileName string) {
+	return info.getBlobFileName() + ".json"
+}
+
+func (info *URLInfo) isEtagValid() bool {
+	if !info.HasEtag {
+		return false
+	}
+
+	var httpClient *http.Client
```
Comment (ybogdanov) [inline src/build/urlfetcher.go]: You can move the fallback to the default http client to the `NewURLFetcherFS` constructor.

File: src/build/copy.go [hunk 3]
```diff
@@ -264,6 +308,24 @@ func listFiles(srcPath string, includes, excludes []string) ([]*uploadFile, erro
 
 	for _, pattern := range includes {
 
+		if isURL(pattern) {
```
Comment (ybogdanov) [inline src/build/copy.go]: Fail if urls are used with `COPY` command.

File: src/build/urlfetcher.go [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+	if err != nil {
+		return false
+	}
+	defer response.Body.Close()
+
+	if response.StatusCode < 200 || 300 <= response.StatusCode {
+		return false
+	}
+
+	etag, ok := response.Header["Etag"]
```
Comment (ybogdanov) [inline src/build/urlfetcher.go]: Please use `response.Header.Get("Etag")`

File: src/build/urlfetcher.go [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+		return err
+	}
+	defer response.Body.Close()
+
+	if response.StatusCode < 200 || 300 <= response.StatusCode {
+		return fmt.Errorf("Got non-2xx status for `%s`: %s", info.URL, response.Status)
+	}
+
+	err = os.MkdirAll(filepath.Dir(info.FileName), 0755)
+	if err != nil {
```
Comment (ybogdanov) [inline src/build/urlfetcher.go]: `if err = os.MkdirAll(filepath.Dir(info.FileName), 0755); err != nil {`

File: src/build/urlfetcher.go [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+}
+
+func (info *URLInfo) load() (ok bool, err error) {
+	fileName := info.getInfoFileName()
+
+	data, err := ioutil.ReadFile(fileName)
+
+	if err != nil {
+		//if pe, ok0 := err.(*os.PathError); ok0 && pe.Err == os.ErrNotExist {
+		if _, ok0 := err.(*os.PathError); ok0 {
```
Comment (ybogdanov) [inline src/build/urlfetcher.go]: `if os.IsNotExist(err) {`

File: src/build/urlfetcher.go [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+	data, err := json.Marshal(info)
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile(fileName, data, 0644)
+}
+
+func (info *URLInfo) dump() (data string, err error) {
+	data0, err := json.Marshal(info)
+	return string(data0), err
```
Comment (ybogdanov) [inline src/build/urlfetcher.go]: You can also use `pretty.Sprint` and/or implement `String` interface for such stuff.

File: src/build/urlfetcher.go [hunk 8]
```diff
@@ -0,0 +1,277 @@
+package build
```
Comment (ybogdanov) [inline src/build/urlfetcher.go]: Insert Grammarly copyright, please.

File: test/addurl_test.go [hunk 9]
```diff
@@ -0,0 +1,429 @@
...
+func TestAddUrl_NoCache(t *testing.T) {
+	tag, cleanup := generateTag("add-url-build-no-cache")
+	defer cleanup()
+
+	sampleDir := makeTmpDir(t, map[string]string{
+		"file1.txt": "content1",
+		"file2.txt": "hello",
+		"file3.txt": "content3",
+	})
+	//defer os.RemoveAll(sampleDir)
```
Comment (ybogdanov) [inline test/addurl_test.go]: uncomment

File: src/build/copy.go [hunk 10]
```diff
@@ -309,6 +309,10 @@ func listFiles(srcPath string, includes, excludes []string, urlFetcher URLFetche
 	for _, pattern := range includes {
 
 		if isURL(pattern) {
+			if cmdName == "COPY" {
+				return nil, fmt.Errorf("can't list downloaded url in COPY command: '%s'", pattern)
```
Comment (ybogdanov) [inline src/build/copy.go]: can't use url in COPY command

File: src/build/copy.go [hunk 11]
```diff
@@ -51,6 +51,49 @@ type uploadFile struct {
...
+		return s, fmt.Errorf("Invalid ADD format - at least two arguments required")
+	}
+
+	var (
+		src  = args[0 : len(args)-1]
+		dest = filepath.FromSlash(args[len(args)-1]) // last one is always the dest
+	)
+
+	// If destination is not a directory (no leading slash)
+	hasLeadingSlash := strings.HasSuffix(dest, string(os.PathSeparator))
```
Comment (romank87) [inline src/build/copy.go]: should be hasTrailingSlash

**Pull request merged**